FEAT: Implement explicit channel semantics modeling

sanjib-sen opened the pull request:
## 🚀 **Pull Request: Implement Explicit Channel Semantics Modeling for AFG**

###  **🆕 New Updates**

- **Improved main function detection algorithm**: made it more dynamic to search for the lang_start call in the first few instructions rather than assuming it's always at position 3. This should make it work across different compilation environments.
- **Switched to function name demangling** for precise detection of channel operations, eliminating reliance on substring matching.
- **Removed false positives**: Non-channel functions (such as `MyOwnStructure::send()`) are no longer incorrectly identified as channel operations.
- **Redesigned channel representation**: Replaced `ChannelEndpoint` with a more robust `ChannelInfo` structure to capture complete channel instances and relationships.
- **Prevented channel duplication**: The analysis now checks for existing `ChannelInfo` instances to ensure no duplicates are created for the same channel creation instruction.
- **Enhanced constraint integration**: Channel constraints are now generated and applied iteratively within the main pointer analysis loop for improved accuracy.
- **Expanded test output and checklist**: Updated documentation, output examples, and verification steps to reflect the new detection methods and improvements.
- **Added support for both `std::sync::mpsc` and `tokio::sync::mpsc`** channel variants.
- **TODO**: Rust based example test needs to be updated

### **📋 Summary**

This PR implements explicit channel semantics modeling in the AFG (Access Flow Guard) pointer analysis framework, addressing the limitation where channels (like `mpsc::Sender` and `mpsc::Receiver`) were treated as regular pointer types without understanding their special communication semantics.

### **🎯 Problem Addressed**

**Before this implementation:**
- Channel operations (`send()`, `recv()`) were treated as regular function calls
- Missing data flow connections between sender and receiver across thread boundaries  
- Incomplete taint propagation through channel communications
- No channel-specific constraints in security analysis
- `parseMainFn` only looks for `lang_start` in the 3rd instruction from the real main function > first argument

**Example of the problem:**
```rust
let (tx, rx) = mpsc::channel();
let tainted_data = get_tainted_value();
tx.send(tainted_data).await.unwrap();    // Send tainted data
let received = rx.recv().await.unwrap(); // Should be tainted but wasn't detected
```

### **✅ Solution Implemented**

**New Components:**
- **`ChannelSemantics` class**: Detects and analyzes channel operations in LLVM IR
- **`ChannelEndpoint` structure**: Represents sender/receiver endpoint relationships
- **`ChannelOperation` structure**: Models CREATE, SEND, and RECV operations
- **Integration with `PointerAnalysis`**: Seamless integration with existing framework

**Key Features:**
- ✅ **Channel Operation Detection**: Identifies `mpsc::channel()`, `send()`, and `recv()` calls
- ✅ **Endpoint Relationship Tracking**: Maps sender/receiver pairs from channel creation
- ✅ **Constraint Generation**: Creates data flow constraints for taint propagation
- ✅ **Inter-thread Analysis**: Enables security analysis across thread boundaries

### **🔧 Technical Implementation**

**Files Added:**
- `include/ChannelSemantics.h` - Channel semantics interface and structures
- `src/ChannelSemantics.cpp` - Channel detection and analysis implementation
- `examples/channel-test/` - Complete test suite with documentation

**Files Modified:**
- `include/PointerAnalysis.h` - Integration points and public interface updates
- `src/PointerAnalysis.cpp` - Channel detection in call handling and constraint integration
- `CMakeLists.txt` - Build configuration for new source files
- `.gitignore` - Proper ignore rules for build artifacts

**Integration Points:**
1. **Detection Phase**: `handleCallInst()` identifies channel operations during analysis
2. **Analysis Phase**: `analyzeChannelOperations()` processes channel relationships  
3. **Constraint Integration**: `integrateChannelConstraints()` applies channel-specific constraints
4. **Re-solving**: Constraints are re-solved with enhanced data flow modeling

### **🧪 Testing & Verification**

**Test Cases Included:**
- ✅ **Manual Channel Test**: Demonstrates channel operation detection with mock LLVM IR
- ✅ **Baseline Test**: Verifies no false positives on non-channel code
- ✅ **Rust Source Example**: Complete Rust program for end-to-end testing

**Verification Results:**
```
=== Channel Semantics Analysis ===
Channel Endpoints (2):
  RECEIVER - Value: %receiver = extractvalue { ptr, ptr } %channel_result, 1
  SENDER - Value: %sender = extractvalue { ptr, ptr } %channel_result, 0

Channel Operations (3):
  CREATE - Instruction: %channel_result = call { ptr, ptr } @_ZN4mpsc7channel...
  SEND - Instruction: %send_result = call i32 @_ZN4mpsc6Sender4send...
  RECV - Instruction: %recv_result = call ptr @_ZN4mpsc8Receiver4recv...

Channel Pairs (1):
  Channel ID: %channel_result = call { ptr, ptr } @_ZN4mpsc7channel...
```

**Screenshot**
<img width="1231" alt="image" src="https://github.com/user-attachments/assets/814289d9-7412-42c8-b31e-c6b0e1b1984b" />


### **📈 Impact & Benefits**

**Security Analysis Enhancements:**
- 🔒 **Enhanced Taint Propagation**: Tainted data sent through channels is now properly tracked
- 🔍 **Inter-thread Data Flow**: Analysis can track data flow across thread boundaries
- 🛡️ **Channel-Specific Policies**: Enables enforcement of access control policies on channels
- 🚨 **Vulnerability Detection**: Better detection of data leaks through channel communications

### **📋 How to Test**

**Build and Test Instructions:**
```bash
# From project root
mkdir build && cd build
cmake -DLLVM_DIR=$(llvm-config --cmakedir) .. && make

# Test channel semantics detection
opt \
  -load-pass-plugin ./libPointerAnalysisPass.so \
  -passes=pointer-analysis \
  -disable-output \
  ../examples/channel-test/channel-test-manual.ll

# Verify baseline (no false positives)
opt \
  -load-pass-plugin ./libPointerAnalysisPass.so \
  -passes=pointer-analysis \
  -disable-output \
  ../examples/demo/demo-r68_llvm17_map.ll

# Rust example test
cd ../examples/channel-test
cargo rustc -- --emit=llvm-ir -C debuginfo=0 -C opt-level=0
cd ../../build
opt -load-pass-plugin ./libPointerAnalysisPass.so \
  -passes=pointer-analysis \
  -disable-output \
  ../examples/channel-test/target/debug/deps/channel_test-*.ll
```

### **📖 Documentation**

Complete technical documentation available in:
- `examples/channel-test/README.md` - Comprehensive implementation guide
- Inline code comments explaining channel detection logic
- Usage examples and expected outputs for verification

april1989 commented on examples/channel-test/README.md:
```diff
@@ -0,0 +1,331 @@
+
+    std::string funcName = func->getName().str();
+
+    // Look for send operation patterns
```
if using `sync::mpsc`, the signature of the sender and receiver function (e.g., `sync::mpsc::Send`) should always be the same (either small cases or capital). the current checking is way to conservative so that a function like `MyOwnStructure::send()` will be identified as a `sync::mpsc`, which is wrong. you can refer to https://github.com/bozhen-liu/AFG/blob/e42f3a8eb9818f320e66edbbc01225b007f491ab/src/OriginPointerAnalysis.cpp#L41 on how to check the function names.

april1989 commented on examples/channel-test/README.md:
```diff
@@ -0,0 +1,331 @@
+The channel semantics are integrated into the main pointer analysis workflow:
+
+1. **Detection Phase**: During `handleCallInst`, channel operations are detected and recorded
+2. **Analysis Phase**: After main pointer analysis, `analyzeChannelOperations()` processes channel relationships
```
it's better to analyze the channel ops and resolve the channel constraints together with the main pointer analysis phase (i.e., solve()). This means that you may need to adjust the two functions (i.e., `analyzeChannelOperations()` and `integrateChannelConstraints()`).

april1989 commented on src/ChannelSemantics.cpp:
```diff
@@ -0,0 +1,240 @@
+    // This is complex in LLVM IR as tuples are often destructured immediately
+    
+    // Create a unique channel ID based on the call instruction
+    Value* channel_id = channel_create;
```
everytime when we see the same `channel_create` instruction, this code is going to create new ChannelEndpoints instead of using the existing one.

april1989 commented on src/ChannelSemantics.cpp:
```diff
@@ -0,0 +1,240 @@
+                        
+                        // Create constraint: op->data_value flows to recv_op->instruction
+                        Node* sendNode = analysis->getOrCreateNode(op->data_value);
+                        Node* recvNode = analysis->getOrCreateNode(recv_op->instruction);
```
what is `recv_op->instruction`?

sanjib-sen replied:
`recv_op->instruction` is the receive operation in the LLVM IR (LLVM CallInst*). For example,

```rs
let received = rx.recv().unwrap();
```

The generated IR Code
```ll
%recv_result = call ptr @_ZN4mpsc8Receiver4recv17h1234567890abcdefE(ptr %receiver)
```

In this case, recv_op->instruction points to this entire call instruction.

In LLVM, the **CallInst*** means  both the instruction and the value it returns

april1989 commented on src/ChannelSemantics.cpp:
```diff
@@ -0,0 +1,240 @@
+    // This is a simplified version - real implementation would need
+    // more sophisticated demangling
+    
+    if (funcName.find("mpsc") != std::string::npos) {
```
the same here, please use exact function signatures, not partial string pattern checking.

april1989 commented on src/PointerAnalysis.cpp:
```diff
@@ -46,6 +46,12 @@ void PointerAnalysis::analyze(Module &M)
 
     errs() << "Pointer analysis completed.\n";
 
+    // Analyze channel operations after main analysis
```
it's better to integrate these two functions into the `solveConstraints()` phase.

april1989 commented on examples/channel-test/README.md:
````diff
@@ -0,0 +1,331 @@
+#### 1. Channel Endpoint Representation (`ChannelEndpoint`)
+
+```cpp
+struct ChannelEndpoint {
````
I don't like the name using `Endpoint`, how about `ChannelInfo` to represent the entire channel instance, including its creation and both endpoints?

april1989 reviewed (commented):
Thanks for the PR! I’ve left several inline comments focused on the design and workflow. Overall this is close, but a few key areas need cleanup.

sanjib-sen commented:
###  **🆕 New Updates**

- **Switched to function name demangling** for precise detection of channel operations, eliminating reliance on substring matching.
- **Removed false positives**: Non-channel functions (such as `MyOwnStructure::send()`) are no longer incorrectly identified as channel operations.
- **Redesigned channel representation**: Replaced `ChannelEndpoint` with a more robust `ChannelInfo` structure to capture complete channel instances and relationships.
- **Prevented channel duplication**: The analysis now checks for existing `ChannelInfo` instances to ensure no duplicates are created for the same channel creation instruction.
- **Enhanced constraint integration**: Channel constraints are now generated and applied iteratively within the main pointer analysis loop for improved accuracy.
- **Expanded test output and checklist**: Updated documentation, output examples, and verification steps to reflect the new detection methods and improvements.
- **Added support for both `std::sync::mpsc` and `tokio::sync::mpsc`** channel variants.

### **🚧 Caution**
- The rust example test needs to be updated to represent both `std::sync::mpsc` and `tokio::sync::mpsc`** channel variants.

bozhen-liu commented on include/ChannelSemantics.h:
```diff
@@ -0,0 +1,93 @@
+    }
+};
+
+// Represents a channel operation (send/recv)
```
// Represents a channel operation (send/recv/creation)

sanjib-sen replied:
Fixed in https://github.com/bozhen-liu/AFG/pull/1/commits/9f6616c74085fbcbcd669cf59a15cd24257d357a

bozhen-liu commented on include/ChannelSemantics.h:
```diff
@@ -0,0 +1,93 @@
+    llvm::Instruction* instruction;  // The call instruction
+    ChannelInfo* channel_info;       // Associated channel instance
+    llvm::Value* data_value;         // Data being sent/received (null for recv)
+    bool is_sender_operation;        // true for send operations, false for recv operations
```
any chance `is_sender_operation == false` includes channel creation? maybe use `operation` for matching the operation type? actually, this `is_sender_operation` is never used anywhere else and `operation` already serves the same purpose, I think you can remove it.

sanjib-sen replied:
Fixed in https://github.com/bozhen-liu/AFG/pull/1/commits/9f6616c74085fbcbcd669cf59a15cd24257d357a

bozhen-liu commented on src/PointerAnalysis.cpp:
```diff
@@ -555,7 +591,33 @@ void PointerAnalysis::visitInvokeInst(InvokeInst &II)
 {
+    Function *caller = CI.getFunction(); // Get the caller function
     Function *calledFn = CI.getCalledFunction();
+    
```
the following code to handle channels inside the if block can be isolated into another function in PointerAnalysis

sanjib-sen replied:
Fixed in https://github.com/bozhen-liu/AFG/pull/1/commits/9ae6640ed009f6c7880ad03e3519b0c0698bb1be

bozhen-liu commented on src/PointerAnalysis.cpp:
```diff
@@ -602,11 +664,28 @@ void PointerAnalysis::visitInstruction(Instruction &I)
 }
 
 void PointerAnalysis::solveConstraints()
+{
```
can you make this the following format:

```
void PointerAnalysis::solveConstraints()
{
    bool changed = true;
    while (changed)
    {
        changed = false;
        for (const auto &constraint : Worklist)
        {
          ... // original code here for Assign, Store, Load
          case Channel:   
             changed = handleChannelConstraints(); // this function handleChannelConstraints includes the code for  analyzeChannelOperations and integrateChannelConstraints
        }
    }
}
```
then let's create a new enumerator in PointerAnalysis.h:
```
    enum ConstraintType
    {
        Assign,
        Load,
        Store,
        Channel // used when considering channel ops 
    };
```

sanjib-sen replied:
Implemented in https://github.com/bozhen-liu/AFG/pull/1/commits/64ae2541ac5628b80daacf1f71189e592b67613f

sanjib-sen replied:
This was a great suggestion. Now all constraints (pointer and channel) are processed in one unified loop. It will be easier to add new constraint types in the future. Also, channel constraints will be benefited from the same fixed-point iteration as other constraints

bozhen-liu commented on include/ChannelSemantics.h:
```diff
@@ -0,0 +1,93 @@
+    llvm::Value* sender_value;      // The sender endpoint value
+    llvm::Value* receiver_value;    // The receiver endpoint value
+    llvm::Type* data_type;          // Type of data being transmitted
+    llvm::Instruction* creation_call; // The channel creation instruction
```
add a new field `Context context` here with default value `Everywhere`, so this code can be easily integrated with `KCallsitePointerAnalysis` and `OriginPointerAnalysis`.

sanjib-sen replied:
Added in https://github.com/bozhen-liu/AFG/pull/1/commits/28f8da365d116b82cfad04becb84c0b3f6d43d8c

bozhen-liu reviewed (commented):
This PR seems great. Only minor issues to resolve. Thanks!

sanjib-sen commented:
Resolved the issues and replied with the commit hashes, thanks.
Your suggestions were top notch for code modularity, and extensibility.

sanjib-sen commented:
Professor Liu,
I resolved the issues and left comments with individual commit hashes that
resolved the particular issue for your convenience.

Thanks,
Sanjib





On Sun, Jun 8, 2025 at 2:43 AM Bozhen Liu ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> This PR seems great. Only minor issues to resolve. Thanks!
> ------------------------------
>
> In include/ChannelSemantics.h
> <https://github.com/bozhen-liu/AFG/pull/1#discussion_r2134119685>:
>
> > +    llvm::Value* receiver_value;    // The receiver endpoint value
> +    llvm::Type* data_type;          // Type of data being transmitted
> +    llvm::Instruction* creation_call; // The channel creation instruction
> +
> +    ChannelInfo(llvm::Value* channel, llvm::Value* sender, llvm::Value* receiver,
> +                llvm::Type* dtype, llvm::Instruction* create_call)
> +        : channel_id(channel), sender_value(sender), receiver_value(receiver),
> +          data_type(dtype), creation_call(create_call) {}
> +
> +    bool operator==(const ChannelInfo& other) const {
> +        return channel_id == other.channel_id && sender_value == other.sender_value &&
> +               receiver_value == other.receiver_value;
> +    }
> +};
> +
> +// Represents a channel operation (send/recv)
>
> // Represents a channel operation (send/recv/creation)
> ------------------------------
>
> In include/ChannelSemantics.h
> <https://github.com/bozhen-liu/AFG/pull/1#discussion_r2134121141>:
>
> > +
> +    bool operator==(const ChannelInfo& other) const {
> +        return channel_id == other.channel_id && sender_value == other.sender_value &&
> +               receiver_value == other.receiver_value;
> +    }
> +};
> +
> +// Represents a channel operation (send/recv)
> +struct ChannelOperation {
> +    enum ChannelOpType { SEND, RECV, CHANNEL_CREATE };
> +
> +    ChannelOpType operation;
> +    llvm::Instruction* instruction;  // The call instruction
> +    ChannelInfo* channel_info;       // Associated channel instance
> +    llvm::Value* data_value;         // Data being sent/received (null for recv)
> +    bool is_sender_operation;        // true for send operations, false for recv operations
>
> any chance is_sender_operation == false includes channel creation? maybe
> use operation for matching the operation type? actually, this
> is_sender_operation is never used anywhere else and operation already
> serves the same purpose, I think you can remove it.
> ------------------------------
>
> In src/PointerAnalysis.cpp
> <https://github.com/bozhen-liu/AFG/pull/1#discussion_r2134137838>:
>
> >      Function *calledFn = CI.getCalledFunction();
> +
>
> the following code to handle channels inside the if block can be isolated
> into another function in PointerAnalysis
> ------------------------------
>
> In src/PointerAnalysis.cpp
> <https://github.com/bozhen-liu/AFG/pull/1#discussion_r2134141069>:
>
> > @@ -602,11 +664,28 @@ void PointerAnalysis::visitInstruction(Instruction &I)
>  }
>
>  void PointerAnalysis::solveConstraints()
> +{
>
> can you make this the following format:
>
> void PointerAnalysis::solveConstraints()
> {
>     bool changed = true;
>     while (changed)
>     {
>         changed = false;
>         for (const auto &constraint : Worklist)
>         {
>           ... // original code here for Assign, Store, Load
>           case Channel:
>              changed = handleChannelConstraints(); // this function handleChannelConstraints includes the code for  analyzeChannelOperations and integrateChannelConstraints
>         }
>     }
> }
>
> then let's create a new enumerator in PointerAnalysis.h:
>
>     enum ConstraintType
>     {
>         Assign,
>         Load,
>         Store,
>         Channel // used when considering channel ops
>     };
>
> ------------------------------
>
> In include/ChannelSemantics.h
> <https://github.com/bozhen-liu/AFG/pull/1#discussion_r2134144988>:
>
> > +#include "llvm/IR/Instructions.h"
> +#include "llvm/Support/raw_ostream.h"
> +#include <unordered_map>
> +#include <unordered_set>
> +#include <vector>
> +#include <utility>
> +
> +namespace llvm {
> +
> +// Represents a complete channel instance including creation and both endpoints
> +struct ChannelInfo {
> +    llvm::Value* channel_id;        // Unique identifier for the channel (creation call)
> +    llvm::Value* sender_value;      // The sender endpoint value
> +    llvm::Value* receiver_value;    // The receiver endpoint value
> +    llvm::Type* data_type;          // Type of data being transmitted
> +    llvm::Instruction* creation_call; // The channel creation instruction
>
> add a new field Context context here with default value Everywhere, so
> this code can be easily integrated with KCallsitePointerAnalysis and
> OriginPointerAnalysis.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bozhen-liu/AFG/pull/1#pullrequestreview-2907735084>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANB5NRV7LALCUOECVBZ6EID3CNFGTAVCNFSM6AAAAAB6BF3HFKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDSMBXG4ZTKMBYGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
